pkg/apis/edge/v1alpha1/customize.go

```diff
@@ -0,0 +1,80 @@
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// ParameterExpansionAnnotationKey, when paired with the value "true" in an annotation of
+// an object subject to edge management, indicates that parameter expansion should be
+// bundled with propagation from center to edge.
+//
+// Parameter expansion applies to every leaf string in the object, and involves
+// two substring replacements.
+// One is replacing "%%" with "%".
+// The other replaces every substring of the form "%(parameter_name)" with the destination's
+// value for the named parameter.  A parameter_name can be any label or annotation key.
+//
+// A destination is a [kcp Location](https://github.com/kcp-dev/kcp/blob/v0.8.2/pkg/apis/scheduling/v1alpha1/types_location.go#L50)
+// and its labels and annotations provide parameter values (with labels taking priority over annotations).
+//
+// Note that this sort of customization has limited applicability.  It can only be used where
+// the un-expanded string passes the validation conditions of the relevant object type.
+// For more broadly applicable customization, see Customizer objects.
+const ParameterExpansionAnnotationKey string = "edge.kcp.io/expand-parameters"
+
+// CustomizerAnnotationKey is the key of an annotation that identifies the Customizer
+// that applies to the annotated object as it propagates from center to edge.
+// When the annotated object is namespaced, the Customizer object to apply is in
+// the same namespace and the value of this annotation is the name of the Customizer object;
+// otherwise the value of this annotation uses the form "namespace/name" to reference
+// the desired Customizer.
+const CustomizerAnnotationKey string = "edge.kcp.io/customizer"
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Customizer defines modifications to make to the relevant objects as they propagate
+// from center to edge.
+//
+// The relevant objects are those with an annotation whose key is
+// "edge.research.ibm.com/customizer" and whose value is the name of this object.
```

Comment:
Domain name update needed here too.
Also, the "whose value is the name of this object" does not cover the case of non-namespaced workload objects, this should be fixed (not necessarily by copying text, referring to it would probably be better).

Reply:
updated comment, inserted compiler directives, included license text, and ran make crds codegen.